src/parser/SGTypes.ts

```diff
@@ -0,0 +1,373 @@
+    }
+
+    constructor(
+        public tag: SGToken,
+        public attributes: SGAttribute[] = [],
+        public range?: Range
+    ) {}
+
+    getSGAttribute(name: string): SGAttribute | undefined {
+        return this.attributes.find(att => att.key.text.toLowerCase() === name);
+    }
+
+    getAttribute(name: string): string | undefined {
+        return this.getSGAttribute(name)?.value?.text;
+    }
+
+    setAttribute(name: string, value: string) {
+        const attr = this.getSGAttribute(name);
+        if (attr) {
+            if (value) {
+                attr.value = { text: value };
+                attr.range = undefined;
+            } else {
+                this.attributes.splice(this.attributes.indexOf(attr));
+            }
+        } else if (value) {
+            this.attributes.push({
+                key: { text: name },
+                value: { text: value }
+            });
+        }
+    }
+
+    transpile(state: SGTranspileState): SourceNode {
+        return new SourceNode(null, null, state.source, [
+            state.indent,
+            '<',
+            state.transpileToken(this.tag),
+            ...state.transpileAttributes(this.attributes),
+            ...this.transpileBody(state)
+        ]);
+    }
+
+    protected transpileBody(state: SGTranspileState): (string | SourceNode)[] {
+        return [' />\n'];
+    }
+}
+
+export class SGProlog extends SGTag {
+
+    transpile(state: SGTranspileState) {
+        return new SourceNode(null, null, state.source, [
+            '<?xml',
+            ...state.transpileAttributes(this.attributes),
+            ' ?>\n'
+        ]);
+    }
+}
+
+export class SGNode extends SGTag {
+
+    constructor(
+        tag: SGToken,
+        attributes?: SGAttribute[],
+        public children: SGNode[] = [],
+        range?: Range
+    ) {
+        super(tag, attributes, range);
+    }
+
+    protected transpileBody(state: SGTranspileState): (string | SourceNode)[] {
+        if (this.children.length > 0) {
+            const body: (string | SourceNode)[] = ['>\n'];
+            state.blockDepth++;
+            body.push(...this.children.map(node => node.transpile(state)));
+            state.blockDepth--;
+            body.push(state.indent, '</', this.tag.text, '>\n');
+            return body;
+        } else {
+            return super.transpileBody(state);
+        }
+    }
+}
+
+export class SGChildren extends SGNode {
+
+    constructor(
+        tag: SGToken = { text: 'children' },
+        children: SGNode[] = [],
+        range?: Range
+    ) {
+        super(tag, [], children, range);
+    }
+}
+
+export class SGScript extends SGTag {
+
+    get type() {
+        return this.getAttribute('type');
+    }
+    set type(value: string) {
+        this.setAttribute('type', value);
+    }
+
+    get uri() {
+        return this.getAttribute('uri');
+    }
+    set uri(value: string) {
+        this.setAttribute('uri', value);
+    }
+
+    constructor(
```

Comment:
Can you lift all of the constructor functions to the top of these classes? I don't have a lint rule for this (is there one?), but I've tried to follow that convention in most of this project.

Reply:
I don't really agree with your pattern but ok.